Added docker-compose.ssl.yaml.example

k-brahma opened the pull request:
# Checklist:
> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.
- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description
This pull request adds a `docker-compose.ssl.yaml.example` file to simplify the SSL certification process for Dify deployments. This addition will help streamline the SSL setup process for Dify users deploying with HTTPS.

## Expected certification process with this file:

1. Change current directory and copy the file:
   ```bash
   # Move to Dify's docker directory
   cd dify/docker
   # Create SSL configuration file for Docker Compose(New!)
   cp docker-compose.ssl.yaml.example docker-compose.ssl.yaml
   ```

2. Install certbot and obtain certification:
   ```bash
   sudo apt update
   sudo apt install certbot
   sudo certbot certonly --standalone -d <your_domain>
   # or use dns challenge
   # sudo certbot certonly --manual --preferred-challenges dns -d <your_domain>
   ```

3. Copy the certificate files to ./nginx/ssl/:
   ```bash
   sudo cp /etc/letsencrypt/live/<your_domain>/fullchain.pem ./nginx/ssl/<your_domain>.crt
   sudo cp /etc/letsencrypt/live/<your_domain>/privkey.pem ./nginx/ssl/<your_domain>.key
   ```

4. Copy .env.example and edit the new .env file:
   ```bash
   cp .env.example .env
   ```
   Then edit the .env file to include:
   ```
   NGINX_HTTPS_ENABLED=true # change from false to true
   NGINX_SSL_PORT=443
   NGINX_SSL_CERT_FILENAME=<your_domain>.crt
   NGINX_SSL_CERT_KEY_FILENAME=<your_domain>.key
   ```

Fixes #[Issue number if applicable]

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions
Please follow these steps to test the changes:

1. Clone the repository and navigate to the docker directory.
2. Copy the `docker-compose.ssl.yaml.example` file to `docker-compose.ssl.yaml`.
3. Follow the steps outlined in the Description section to set up SSL for a test domain.
4. Verify that the Dify application can be accessed via HTTPS after completing the setup.

- [x] Test on a local development environment
- [x] Test on a staging server (if applicable)

k-brahma committed:
Added docker-compose.ssl.yaml.example

k-brahma committed:
SSL setup guide

k-brahma commented:
Update: detaild SSL setup manual added.

https://github.com/langgenius/dify/pull/6509/commits/0aa0c8fe9ebcccfac513320714a6afe40ca4aa73

k-brahma committed:
replaced strings: ~/dify -> /path/to/dify

k-brahma commented:
# Added Comprehensive SSL Setup Guide for Dify Deployment

## PR Overview

This PR adds a detailed SSL configuration guide (`SSL_SETUP.md`) to assist users in setting up HTTPS for their Dify deployments. The guide is based on practical experience and has been thoroughly tested. It also leverages the docker-compose.ssl.yaml.example file, which significantly simplifies the SSL setup process.

## Key Features

1. Step-by-step instructions for obtaining and installing SSL certificates

2. Clear explanations for editing Dify configuration files

3. Verified distribution list with status indicators

4. Optional automatic certificate renewal script

5. Utilization of docker-compose.ssl.yaml.example for easy SSL configuration

## Why This is Valuable

- Simplifies the SSL setup process for users of all experience levels

- Enhances security for Dify deployments by encouraging HTTPS usage

- Provides a tested and verified approach, reducing potential issues

- Encourages community contributions through the distribution status table

- Dramatically simplifies SSL setup with the use of docker-compose.ssl.yaml.example

  - Users can set up SSL with a simple cp command, saving time and reducing errors

## Testing and Verification

- Successfully tested on Ubuntu 24.04 LTS (x86_64)

- Includes notes on potential compatibility with other Debian-based systems

## Additional Notes

- The guide is designed to be easily expandable as more distributions are tested

- Welcomes community feedback and contributions for other distributions

- The docker-compose.ssl.yaml.example file is a key component that makes the entire process much more user-friendly

I believe this guide, along with the docker-compose.ssl.yaml.example file, will significantly improve the user experience for Dify adopters, especially those new to SSL configuration. It aligns with the project's goal of making advanced AI applications accessible, secure, and easy to set up.

I'm waiting for your feedback and suggestions for improvement.

takatost commented on docker/docker-compose.ssl.yaml.example:
```diff
@@ -0,0 +1,14 @@
+
+version: '3'
+services:
+  nginx:
```
May I ask what nginx is used for here?

takatost commented on docker/docker-compose.ssl.yaml.example:
Can this file be merged with the main docker compose file? You can add `certbot` service to `/path/to/dify/docker/docker-compose.yaml` and configure it with the `profiles` parameter, and if you need to use certbot at startup, you can start it with the following command:

``
docker-compose --profile certbot up
``

Or you can modify the `.env` to achieve the same effect.

``
COMPOSE_PROFILES=${VECTOR_STORE:-weaviate},certbot
``

k-brahma commented:
This PR has been superseded by #6702, which implements a Docker-based solution for SSL certificate management using Certbot. The new PR addresses the feedback received on this PR and provides a more integrated approach aligned with the project's containerized architecture.

Key improvements in #6702:
1. Simplified SSL certificate acquisition and renewal process
2. Improved security through automated certificate management
3. Maintained backward compatibility for existing setups

As a result, I'm closing this PR in favor of the new implementation. Please refer to #6702 for the updated SSL setup process. 
@takatost Thank you for your feedback and guidance, which led to this improved solution. Your input was invaluable in shaping this new approach.
